src/Microsoft.Extensions.Primitives/StringSegment.cs

```diff
@@ -361,21 +319,12 @@ public string Substring(int offset)
         /// <param name="length">The number of characters in the substring.</param>
         /// <returns>A <see cref="string"/> that is equivalent to the substring of length <paramref name="length"/> that begins at
         /// <paramref name="offset"/> in this <see cref="StringSegment"/></returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public string Substring(int offset, int length)
         {
-            if (!HasValue)
+            if (!HasValue || offset < 0 || offset + length > Length || length < 0)
```

Comment:
Misses the `Offset + offset + length > Buffer.Length` test?

Reply:
Know this is same as before; but `offset + length` could overflow and become negative; therefore passing the `> Length` test; so you probably want to use subtraction and go via `uint` to move negative values to very high values; which will also cut down the number of tests e.g.
```csharp
if (!HasValue || 
    (uint)offset > (uint)(buffer.Length - offset) || 
    (uint)length > (uint)(buffer.Length - offset - Offset))
{
    ThrowInvalidArguments(offset, length);
}
```

Reply:
The overflow-catch is good -- I'll add this one.

But is the check with `Buffer.Length` relevant?
`SubString` should only be allowed to get the chars in this _StringSegment_ -- i.e. the chars in the range [`Offset`, `Offset + Length`) -- and should not be allowed to read beyond `Offset + Length`. Hence the check for `> Buffer.Length` is not necessary.

From the doc (xml comment) for `Substring`:
> A `string` that is equivalent to the substring of length `length` that begins at `offset` in this `StringSegment`.

It depends how "in this StringSegment" is interpreted.
Either it means that `offset` is in this StringSegment (and reading beyond StringSegment's boundaries is permitted) or the resulting substring must be within the boundaries of StringSegment.
For me it is the latter. If I have misread/-interpreted it, so correct me please. In this case I'll make the xml-comment more precise.
For me it feels strange when I get a StringSegment passed from which I don't know the buffer and I'm allowed to get a substring up to the end of buffer. More natural seems that I can only get a substring from within the bounds of the given StringSegment.

So the relevant checks are:
* StringSegment must have a value
* offset must be greater than or equal to zero
* length must be greater than or equal to zero
* offset + length must not overflow
* offset + length must be within the bounds of StringSegment, i.e. lower than or equal to Length

So this is
```c#
if (!HasValue || offset < 0 || length < 0 || (uint)(offset + length) > (uint)Length)
{
    ThrowInvalidArguments(offset, length);
}
```

Where one `< 0` check can be eliminated by moving to subtraction from Length. But it seems that the `sub` is slower than the additional check:
``` ini

BenchmarkDotNet=v0.10.12, OS=Windows 10 Redstone 3 [1709, Fall Creators Update] (10.0.16299.309)
Intel Core i7-7700HQ CPU 2.80GHz (Kaby Lake), 1 CPU, 8 logical cores and 4 physical cores
Frequency=2742189 Hz, Resolution=364.6722 ns, Timer=TSC
.NET Core SDK=2.1.100
  [Host]     : .NET Core 2.0.5 (Framework 4.6.26020.03), 64bit RyuJIT
  DefaultJob : .NET Core 2.0.5 (Framework 4.6.26020.03), 64bit RyuJIT


```
| Method |     Mean |     Error |    StdDev | Scaled | ScaledSD |
|------- |---------:|----------:|----------:|-------:|---------:|
|   Base | 10.82 ns | 0.1338 ns | 0.1252 ns |   1.00 |     0.00 |
| Test with sub | 11.09 ns | 0.1315 ns | 0.1230 ns |   1.03 |     0.02 |

Base:
```asm
G_M15034_IG02:
       48837DF000           cmp      gword ptr [rbp-10H], 0
       7429                 je       SHORT G_M15034_IG05
       85D2                 test     edx, edx
       7C25                 jl       SHORT G_M15034_IG05
       85C0                 test     eax, eax
       7C21                 jl       SHORT G_M15034_IG05
       8D3402               lea      esi, [rdx+rax]
       3B75FC               cmp      esi, dword ptr [rbp-04H]
       7719                 ja       SHORT G_M15034_IG05
```

Test with sub:
```asm
G_M15034_IG02:
       48837DF000           cmp      gword ptr [rbp-10H], 0
       7426                 je       SHORT G_M15034_IG05
       8B75FC               mov      esi, dword ptr [rbp-04H]
       2BF0                 sub      esi, eax
       3BF2                 cmp      esi, edx
       721D                 jb       SHORT G_M15034_IG05
       85C0                 test     eax, eax
       7C19                 jl       SHORT G_M15034_IG05
```